zulipterminal/core.py

```diff
@@ -262,6 +262,15 @@ def _finalize_show(self, w_list: List[Any]) -> None:
         if focus_position >= 0 and focus_position < len(w_list):
             self.model.msg_list.set_focus(focus_position)
 
+    def exit_with(self, exception_to_raise: Any,
+                  error: Any=None) -> None:
+        if exception_to_raise is None:
+            raise urwid.ExitMainLoop()
+        elif error is None:
+            raise exception_to_raise()
+        else:
+            raise exception_to_raise(error)
+
```

Comment:
Since every time an we have to close ZT when an exception is raised we also need to set `self.ui_loading = False`. This function helps us not repeat `self.ui_loading = False` everywhere. Also, it stops us from using `urwid` in model.py which you were strongly against.

Reply:
I think the spinning cursor is better suited in controller since it marks the beginning of ZT really and not UI.

As for the `ui_loading` I think we can do that.

Reply:
@neiljp  I think perhaps this is what you meant by catching the exception in `core.py` right?

Reply:
Hmm, I agree that the spinning cursor is not part of the UI and that the controller handles things that aren't UI-related, but it also writes to the terminal rather than the UI, and to me that seems within the remit of `run.py` - opinions may differ :)

I'm still not convinced what we gain by using `exit_with`; I would split that out and just leave the `ui_loading` changes in this commit, and put the rest into another commit/branch for now? We can discuss on czo about it?

For the "catching the exception in `core.py`", my recollection was that a solution might be to wrap some code in  `Controller.__init__` inside a `try`, and then on noticing an exception adjust `ui_loading` and re-raise the exception so that `run.py` can handle it.